apps/snfoundry/contracts/src/Marketplace.cairo

```diff
@@ -0,0 +1,369 @@
+            let contract_address = get_contract_address();
+
+            // Check if the buyer has enough funds to buy all the products
+            let mut token_idx = 0;
+            let mut total_price = 0_u256;
+            loop {
+                if token_idx == token_ids.len() {
+                    break;
+                }
+                let stock = self.listed_product_stock.read(*token_ids.at(token_idx));
+                assert(stock > 0, 'Product not available');
+                let token_amount = *token_amount.at(token_idx);
+                assert(stock >= token_amount, 'Not enough stock');
+                total_price += self.listed_product_price.read(*token_ids.at(token_idx)) * token_amount;
+                token_idx += 1;
+            };
+
+            // Transfer the funds
+            total_price += self.calculate_fee(total_price, self.market_fee.read());
+            let strk_token_dispatcher = IERC20Dispatcher {
+                contract_address: contract_address_const::<STRK_TOKEN_ADDRESS>()
+            };
+            assert(strk_token_dispatcher.balance_of(buyer) >= total_price, 'insufficient funds');
+            strk_token_dispatcher.transfer_from(buyer, contract_address, total_price);
+
+            // Transfer the nft products
+            let cofi_collection = ICofiCollectionDispatcher {contract_address:  self.cofi_collection_address.read()};
+            cofi_collection.safe_batch_transfer_from(
+                contract_address, buyer, token_ids, token_amount, array![0].span()
+            );
+
+            // Update stock for products
+            let mut token_idx = 0;
+            loop {
+                if token_idx == token_ids.len() {
+                    break;
+                }
+                let stock = self.listed_product_stock.read(*token_ids.at(token_idx));
+                let token_amount = *token_amount.at(token_idx);
+                let new_stock = stock - token_amount;
+                self.update_stock(*token_ids.at(token_idx), new_stock);
+                token_idx += 1;
+            };
+
+            // TODO: change transfer to use calculate_fee function and add it to the claim balances accordingly
+            if (!self.accesscontrol.has_role(CONSUMER, buyer)) {
+                self.accesscontrol._grant_role(CONSUMER, buyer);
+            }
+        }
+    
+        fn create_product(ref self: ContractState, token_id: u256, initial_stock: u256, price: u256) {
+            self.accesscontrol.assert_only_role(PRODUCER);
+            let cofi_collection = ICofiCollectionDispatcher {contract_address:  self.cofi_collection_address.read()};
+            cofi_collection.mint(get_contract_address(), token_id, initial_stock, array![0].span());
+            
+            let producer = get_caller_address();
+            self.seller_products.write(token_id, producer);
+            self.initialize_product(token_id, producer, initial_stock, price);
+        }
+
+        fn create_products(
```

Comment:
@Jonatan-Chaverri just in case this is missing the metadata.

Reply:
As discussed separately, we would set the base URI for metadata in the constructor and then the ERC-1155 would take to automatically handle the locations of each token specific metadata.